pygen/pygen_src/riscv_directed_instr_lib.py

```diff
@@ -161,3 +163,134 @@ def post_randomize(self):
             instr = self.randomize_gpr(instr)
             self.instr_list.append(instr)
         super().post_randomize()
+
+
+# Push Stack Instructions
+class riscv_push_stack_instr(riscv_rand_instr_stream):
+
+    def __init__(self):
+        super().__init__()
+        self.stack_len = 0
+        self.num_of_reg_to_save = 0
+        self.num_of_redudant_instr = 0
+        self.push_stack_instr = []
+        self.saved_regs = []
+        self.branch_instr = riscv_instr()
+        self.enable_branch = vsc.rand_bit_t()
+        self.push_start_label = ''
+
+    def init(self):
+        self.reserved_rd = [cfg.ra]
+        self.saved_regs = [cfg.ra]
+        self.num_of_reg_to_save = len(self.saved_regs)
+        if self.num_of_reg_to_save * (rcs.XLEN / 8) > self.stack_len:
+            logging.error('stack len [%0d] is not enough to store %d regs',
+                          self.stack_len, self.num_of_reg_to_save)
+        self.num_of_redudant_instr = random.randrange(3, 10)
+        self.initialize_instr_list(self.num_of_redudant_instr)
+
+    def gen_push_stack_instr(self, stack_len, allow_branch=1):
+        self.stack_len = stack_len
+        self.init()
+        self.push_stack_instr = [0] * (self.num_of_reg_to_save + 1)
+        for i in range(len(self.push_stack_instr)):
+            self.push_stack_instr[i] = riscv_instr()
+        self.push_stack_instr[0] = \
+            riscv_instr_ins.get_instr(riscv_instr_name_t.ADDI.name)
+        with self.push_stack_instr[0].randomize_with() as it:
+            self.push_stack_instr[0].rd == cfg.sp
+            self.push_stack_instr[0].rs1 == cfg.sp
+            self.push_stack_instr[0].imm == cfg.stack_len + 1
```

Comment:
`cfg.stack_len` should be inverted first, as seen in the [SV side](https://github.com/google/riscv-dv/blob/master/src/riscv_directed_instr_lib.sv#L319)